library/spark/spark-accumulo-library/src/main/java/uk/gov/gchq/gaffer/sparkaccumulo/operation/rfilereaderrdd/RFileReaderIterator.java

```diff
@@ -0,0 +1,217 @@
+import org.apache.accumulo.core.conf.AccumuloConfiguration;
+import org.apache.accumulo.core.conf.DefaultConfiguration;
+import org.apache.accumulo.core.conf.SiteConfiguration;
+import org.apache.accumulo.core.data.ArrayByteSequence;
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.file.blockfile.impl.CachableBlockFile;
+import org.apache.accumulo.core.file.rfile.RFile;
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.IteratorUtil;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.accumulo.core.iterators.system.MultiIterator;
+import org.apache.accumulo.core.iterators.system.VisibilityFilter;
+import org.apache.accumulo.core.security.Authorizations;
+import org.apache.accumulo.core.util.Pair;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.spark.Partition;
+import org.apache.spark.TaskContext;
+import org.apache.spark.util.TaskCompletionListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A <code>RFileReaderIterator</code> is a {@link java.util.Iterator} formed by merging iterators over
+ * a set of RFiles.
+ */
+public class RFileReaderIterator implements java.util.Iterator<Map.Entry<Key, Value>> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RFileReaderIterator.class);
+    private final Partition partition;
+    private final TaskContext taskContext;
+    private final List<SortedKeyValueIterator<Key, Value>> iterators = new ArrayList<>();
+    private SortedKeyValueIterator<Key, Value> mergedIterator = null;
+    private SortedKeyValueIterator<Key, Value> iteratorAfterIterators = null;
+    private Configuration configuration;
+    private Set<String> auths;
+
+    public RFileReaderIterator(final Partition partition,
+                               final TaskContext taskContext,
+                               final Configuration configuration,
+                               final Set<String> auths) {
+        this.partition = partition;
+        this.taskContext = taskContext;
+        this.configuration = configuration;
+        this.auths = auths;
+        try {
+            init();
+        } catch (final IOException e) {
+            throw new RuntimeException("IOException initialising RFileReaderIterator", e);
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        return iteratorAfterIterators.hasTop();
```

Comment:
Do you need to check for deletions?

From the SortedKeyValueIterator javadoc:

_Note that in minor compaction scope and in non-full major compaction scopes the iterator may see deletion entries._

Reply:
Debateable. We don't provide any way for users to delete data (removal of data that is invalid isn't deletion of data in the sense the javadoc means). But they could manually delete table from the table. I'll add it in to be safe.

Reply:
OK fair enough, if it is a big change just create a separate issue for and it and we can do it later.

Reply:
Separate issue opened: #1248